README
README for Perl extension File-Copy-Recursive-Reduced
  File-Copy-Recursive-Reduced is a Perl library which provides
  subroutines exported on request which are intended to serve as
  drop-in replacements for certain subroutines found in CPAN
  distribution File-Copy-Recursive
  <http://search.cpan.org/~dmuey/File-Copy-Recursive-0.40/> (FCR).

 What Problem Does This Library Address?
  As of early April 2018, FCR was failing to install on several
  important operating systems including FreeBSD and Windows. As a
  consequence, other CPAN distributions which had a direct or
  indirect dependency on FCR (*e.g.*, CPAN-Reporter, DateTime,
  Dist-Zilla) were failing to install on those platforms as well.

  Such failures were "silent" in the sense that CPANtesters
  installations generate no report authors when a distribution's
  upstream dependencies fail to install. No report to the
  CPANtesters database means no email notification to CPAN
  authors. So many authors were unaware that their distributions
  were failing to install on important platforms.

  Uncertainty as to when FCR would be fixed led to the creation of
  File-Copy-Recursive-Reduced (FCR2). Certain CPAN distributions
  adopted it in place of FCR. With the April 19 2018 release of
  FCR version 0.41, the problems that motivated the creation of
  FCR2 have been addressed.

  FCR2 is now feature-complete. It exports three functions on
  demand which are substantially equivalent to their FCR
  equivalents:

  *   "fcopy()"

  *   "dircopy()"

  *   "rcopy()"

  After installation, call perldoc File::Copy::Recursive::Reduced
  for more usage details.

  These functions are quite appropriate in situations such as test
  suites where the user has full knowledge of the files,
  directories and symlinks to be recursively copied and does not
  need to manipulate the environment by setting localized versions
  of FCR's package global variables.

 Install
  This library can be installed in the customary way, *i.e.,* by a
  CPAN installer program such as cpan or cpanm or by satisfying
  its prerequisites (Capture::Tiny, for the test
  suite only) and then calling:

      perl Makefile.PL
      make
      make test
      make install

  For further information, after installation call "perldoc
  File::Copy::Recursive::Reduced".



Changes
Revision history for Perl module File::Copy::Recursive::Reduced

0.008 TBD
    - Encoding directive added to POD, contributed by gregor herrmann from
      Debian.

0.007 Sat Sep 16 07:48:17 AM EDT 2023
    - Corrections (in test files only) to better accommodate Windows path
      separators, mostly provided by Håkon Hægland.

0.006 Fri Apr 20 10:21:48 EDT 2018
    - File::Copy::Recursive 0.41 has been released to CPAN and
      addresses the problem which was the focus of
      File::Copy::Recursive::Reduced.  Hence, FCR2 is now
      feature-complete.

0.005 Thu Apr 19 21:03:28 EDT 2018
    - Implement copying of symlinks within all three public functions.
      At this point we should be able to make substitutions for
      File::Copy::Recursive's fcopy(), dircopy() and rcopy() functions
      in a large proportion of toolchain libraries, particularly in
      test suites.

0.004 Thu Apr 19 11:20:52 EDT 2018
    - Introduce rcopy(), a stripped-down replacement for
      File::Copy::Recursive::rcopy().  (But no support for symlinks
      yet in this or any other exportable function.)

0.003 Tue Apr 17 09:14:19 EDT 2018
    - No changes in functionality; changes only to documentation.

0.002 Mon Apr 16 17:51:13 EDT 2018
    - Object-oriented interface replaced; module now exports
      fcopy() and dircopy() on demand.

    - Should be usable as replacement for File::Copy::Recursive
      in the test suite for CPAN-Reporter.

0.001 Tue Apr 10 22:03:22 2018
    - Original version; created by ExtUtils::ModuleMaker 0.56;
    - Not yet ready for use.




t/lib/Helper.pm
package Helper;
use strict;
use warnings;

our (@EXPORT_OK, @ISA);
use Exporter ();
@ISA = 'Exporter';
@EXPORT_OK = ( qw|
    create_tfile
    create_tfile_and_name_for_new_file_in_same_dir
    get_mode
    create_tsubdir
    get_fresh_tmp_dir
    touch_a_file_and_test
    touch_directories_and_test
    touch_left_path_and_test
    prepare_left_side_directories
    make_mixed_directory
    make_imperfect_mixed_directory
    run_system_cmd
    run_bash_cmd
| );
use File::Basename ( qw| basename dirname | );
use File::Path ( qw| mkpath | );
use File::Spec;
use File::Temp ( qw| tempdir | );
use Path::Tiny;

sub create_tfile {
    my $tdir = shift;
    my $filename = shift || 'old';
    my $f = File::Spec->catfile($tdir, $filename);
    open my $OUT, '>', $f or die "Unable to open for writing: $!";
    binmode $OUT;
    print $OUT "\n";
    close $OUT or die "Unable to close after writing: $!";
    return $f;
}

sub create_tfile_and_name_for_new_file_in_same_dir {
    my $tdir = shift;
    my $new_filename = shift || 'new_file';
    my $old = create_tfile($tdir);
    my $new = File::Spec->catfile($tdir, $new_filename);
    return ($old, $new);
}

sub get_mode {
    my $file = shift;
    return sprintf("%04o" => ((stat($file))[2] & 07777));
}

sub create_tsubdir {
    my $tdir = shift;
    my $old = File::Spec->catdir($tdir, 'old_dir');
    my $rv = mkdir($old);
    die "Unable to create temporary subdirectory for testing: $!"
        unless $rv;
    return $old;
}

sub get_fresh_tmp_dir {
    # Adapted from FCR t/01.legacy.t
    my $tmpd = shift || tempdir( CLEANUP => 1 );
    for my $dir ( _get_dirs($tmpd) ) {
        my @created = mkpath($dir, { mode => 0711 });
        die "Unable to create directory $dir for testing: $!" unless @created;

        path("$dir/empty")->spew("");
        path("$dir/data")->spew("oh hai\n$dir");
        path("$dir/data_tnl")->spew("oh hai\n$dir\n");
        no warnings 'once';
        if ($File::Copy::Recursive::Reduced::CopyLink) {
            symlink( "data",    "$dir/symlink" );
            symlink( "noexist", "$dir/symlink-broken" );
            symlink( "..",      "$dir/symlink-loopy" );
        }
        use warnings;
    }
    return $tmpd;
}

sub _get_dirs {
    # Adapted from FCR t/01.legacy.t
    my $tempd = shift;
    my @dirs = (
        [ qw| orig | ],
        [ qw| orig foo | ],
        [ qw| orig foo bar | ],
        [ qw| orig foo baz | ],
        [ qw| orig foo bar bletch | ],
    );
    my @catdirs = ();
    for my $set (@dirs) {
        push @catdirs, File::Spec->catdir($tempd, @{$set});
    }
    return @catdirs;
}

sub touch_a_file_and_test {
    my $f = shift;
    open my $OUT, '>', $f or die "Unable to open $f for writing";
    print $OUT "\n";
    close $OUT or die "Unable to close $f after writing";
    Test::More::ok(-f $f, "Created $f for testing");
    return 1;
}

sub touch_directories_and_test {
    my ($topdir, $tdir_names) = @_;
    my @tdirs = ();
    for my $d (@{$tdir_names}) {
        my $s = File::Spec->catdir($topdir, $d);
        mkpath($s) or die "Unable to mkpath $s: $!";
        Test::More::ok(-d $s, "Directory $s created");
        push @tdirs, $s;
    }
    return @tdirs;
}

sub touch_left_path_and_test {
    my @dirs = @_;
    my $ldir = File::Spec->catdir(@dirs);
    mkpath($ldir) or die "Unable to mkpath $ldir: $!";
    Test::More::ok(-d $ldir, "Directory $ldir created");
    return $ldir;
}

sub prepare_left_side_directories {
    my ($topdir, $dirname, $subdirs) = @_;
    my $tdir = File::Spec->catdir($topdir, $dirname);
    mkpath($tdir) or die "Unable to mkpath $tdir";
    Test::More::ok(-d $tdir, "Directory $tdir created");
    my $old        = File::Spec->catdir($tdir);
    my $oldtree    = File::Spec->catdir($tdir, @{$subdirs});
    my @created = mkpath($oldtree, { mode => 0711 });
    die "Unable to create directory $oldtree for testing: $!" unless -d $oldtree;
    Test::More::ok(-d $oldtree, "Directory $oldtree created for testing");
    return ($old, $oldtree);
}

sub make_mixed_directory {
    my $topdir = shift;
    my @dirnames = qw( alpha beta gamma );
    my @subdirs = ();
    for my $d (@dirnames) {
        my $p = File::Spec->catdir($topdir, $d);
        mkpath $p or die "Unable to mkpath $p";
        push @subdirs, $p;
    }
    my @next_dirnames = qw( albemarle beverly );
    my @nextdirs = ();
    my @files_created = ();
    for my $d (@subdirs) {
        for my $e (@next_dirnames) {
            my $p = File::Spec->catdir($d, $e);
            mkpath $p or die "Unable to mkpath $p";
            push @nextdirs, $p;
            my @fnames = qw( f1 f2 );
            if ($e eq $next_dirnames[0]) {
                my $f1 = File::Spec->catfile($p, $fnames[0]);
                my $f2 = File::Spec->catfile($p, $fnames[1]);
                for my $f ($f1, $f2) {
                    open my $OUT, '>', $f or die "Unable to open for writing";
                    print $OUT "\n";
                    close $OUT or die "Unable to close after writing";
                    push @files_created, $f;
                }
            }
        }
    }
    my @symlinks_created = ();
    for my $f (@files_created) {
        my $base = basename($f);
        my $dirs = dirname($f);
        if ($base eq 'f1') {
            my @dirs = File::Spec->splitdir($dirs);
            my $orig = File::Spec->catfile($f);
            my $newdir = File::Spec->catdir(@dirs[0 .. ($#dirs -1)], 'beverly');
            my $symlink = File::Spec->catfile($newdir, 'l1');
            symlink $orig, $symlink or die "Unable to symlink";
            push @symlinks_created, $symlink;
        }

    }
    my @dirs_created = (@subdirs, @nextdirs);
    my $rv = {
        dirs => \@dirs_created,
        files => \@files_created,
        symlinks => \@symlinks_created,
    };
    #require Data::Dump;
    #Data::Dump::dd($rv);
    return $rv;
}

sub make_imperfect_mixed_directory {
    my $topdir = shift;
    my @dirnames = qw( alpha beta gamma );
    my @subdirs = ();
    for my $d (@dirnames) {
        my $p = File::Spec->catdir($topdir, $d);
        mkpath $p or die "Unable to mkpath $p";
        push @subdirs, $p;
    }
    my @next_dirnames = qw( albemarle beverly );
    my @nextdirs = ();
    my %files_created = ();
    for my $d (@subdirs) {
        for my $e (@next_dirnames) {
            my $p = File::Spec->catdir($d, $e);
            mkpath $p or die "Unable to mkpath $p";
            push @nextdirs, $p;
            my @fnames = qw( f1 f2 );
            if ($e eq $next_dirnames[0]) {
                my $f1 = File::Spec->catfile($p, $fnames[0]);
                my $f2 = File::Spec->catfile($p, $fnames[1]);
                for my $f ($f1, $f2) {
                    open my $OUT, '>', $f or die "Unable to open for writing";
                    print $OUT "\n";
                    close $OUT or die "Unable to close after writing";
                    #push @files_created, $f;
                    $files_created{$f}++;
                }
            }
        }
    }
    my @symlinks_created = ();
    for my $f (keys %files_created) {
        my $base = basename($f);
        my $dirs = dirname($f);
        if ($base eq 'f1') {
            my @dirs = File::Spec->splitdir($dirs);
            my $orig = File::Spec->catfile($f);
            my $newdir = File::Spec->catdir(@dirs[0 .. ($#dirs -1)], 'beverly');
            my $symlink = File::Spec->catfile($newdir, 'l1');
            symlink $orig, $symlink or die "Unable to symlink";
            push @symlinks_created, $symlink;
        }

    }
    # delete 1 file which is the target of a symlink
    # "/tmp/Gi_iWAqYXz/old/alpha/albemarle/f1"
    for my $f (keys %files_created) {
        my $g = File::Spec->catfile($topdir, $dirnames[0], $next_dirnames[0], 'f1');
        if ($f =~ m{\Q$g\E$}) {
            unlink $g or die "Unable to unlink $g";
            delete $files_created{$f};
            last;
        }
    }
    my @dirs_created = (@subdirs, @nextdirs);
    my $rv = {
        dirs => \@dirs_created,
        files => [ keys %files_created ],
        symlinks => \@symlinks_created,
    };
    return $rv;
}
sub run_bash_cmd {
    my $cmd = shift;
    return run_system_cmd(['bash', '-c', $cmd]);
}

sub run_system_cmd {
    my ($cmd) = @_;

    my $res = system @$cmd;
    if ( $res == -1 ) {
        die "failed to run command @$cmd: $!";
    }
    elsif ($res & 127) {
        die sprintf "system(@$cmd) : killed by signal %d, %s coredump",
            ($res & 127),  ($res & 128) ? 'with' : 'without';
    }
    return $res >> 8;
}

1;


t/lib/CygwinHelper.pm
package CygwinHelper;
use strict;
use warnings;

our (@EXPORT_OK, @ISA);
use Exporter ();
@ISA = 'Exporter';
@EXPORT_OK = qw( env_prefix );

sub env_prefix {
    my $mode = shift;
    my $OS = uc $^O;
    my $prefix = "$OS=";
    $prefix .= "winsymlinks:$mode" if defined $mode;
    return $prefix;
}

1;

t/scripts/cygwin/create_symlink.pl
use strict;
use warnings;
use feature qw(say);

use File::Temp qw(tempdir);

{
    my $dir = tempdir( CLEANUP => 1 );
    chdir $dir;
    my $fn = 'a';
    # Create an empty file..
    open my $fh, '>', $fn or exit 2;
    close $fh or exit 2;
    if( symlink 'a', 'b') {
        exit 0;
    }
    else {
        exit 1;
    }
}

Makefile.PL
use ExtUtils::MakeMaker;
require 5.008;

my $mm_ver = ExtUtils::MakeMaker->VERSION;

WriteMakefile(
    NAME         => 'File::Copy::Recursive::Reduced',
    VERSION_FROM => 'lib/File/Copy/Recursive/Reduced.pm', # finds \$VERSION
    AUTHOR       => 'James E Keenan (jkeenan@cpan.org)',
    ABSTRACT     => 'Recursive copying of files and directories within Perl 5 toolchain',
    MIN_PERL_VERSION => '5.8.0',
    # Modules needed to compile, build or actually run this distribution:
    PREREQ_PM    => {
                        'File::Copy'    => 0,
                        'File::Find'    => 0,
                        'File::Path'    => 0,
                        'File::Spec'    => 0,
                    },
    # Modern EU::MM enables us to identify modules required only for the test
    # suite.
    ( $mm_ver < 6.63_03 ? 'BUILD_REQUIRES' : 'TEST_REQUIRES' ) =>  {
                        'Test::Simple'  => 0.44,
                        'Capture::Tiny' => 0,
                        'File::Path'    => 0,
                        'File::Spec'    => 0,
                        'File::Temp'    => 0,
                    },
    # Modern EU::MM enables 'make dist' to write good META.json (and META.yml)
    ($mm_ver < 6.46 ? () : (META_MERGE => {
        'meta-spec' => { version => 2 },
        dynamic_config => 1,
        resources => {
            homepage => 'http://thenceforward.net/perl/modules/File-Copy-Recursive-Reduced/',
            repository => {
                url => 'https://github.com/jkeenan/file-copy-recursive-reduced.git',
                web => 'https://github.com/jkeenan/file-copy-recursive-reduced',
                type => 'git',
            },
            bugtracker => {
                web => 'https://rt.cpan.org/Public/Dist/Display.html?Name=File-Copy-Recursive-Reduced',
            },
        },
    })),
    ($mm_ver < 6.31 ? () : (LICENSE => 'perl_5')),
);


lib/File/Copy/Recursive/Reduced.pm
package File::Copy::Recursive::Reduced;
use strict;
use warnings;

use parent qw( Exporter );
our @EXPORT_OK = qw( dircopy fcopy rcopy );
our $VERSION = '0.007';

use File::Copy;
use File::Find;
use File::Path qw( mkpath );
use File::Spec;

our $Link       = eval { local $SIG{'__DIE__'}; link    '', ''; 1 } || 0;
our $CopyLink   = eval { local $SIG{'__DIE__'}; symlink '', ''; 1 } || 0;
our $DirPerms   = 0777;

=encoding UTF-8

=head1 NAME

File::Copy::Recursive::Reduced - Recursive copying of files and directories within Perl 5 toolchain

=head1 SYNOPSIS

    use File::Copy::Recursive::Reduced qw(fcopy dircopy);

    fcopy($orig,$new) or die $!;

    dircopy($orig,$new) or die $!;

=head1 DESCRIPTION

This library is intended as a not-quite-drop-in replacement for certain
functionality provided by L<CPAN distribution
File-Copy-Recursive|http://search.cpan.org/dist/File-Copy-Recursive/>.  The
library provides methods similar enough to that distribution's C<fcopy()>,
C<dircopy()> and C<rcopy()> functions to be usable in those CPAN distributions
often described as being part of the Perl toolchain.

=head2 Rationale

F<File::Copy::Recursive> (hereinafter referred to as B<FCR>) is heavily used
in other CPAN libraries.  Out of over 30,000 other CPAN distributions studied
in early 2018, it ranks by one calculation as the 129th highest distribution
in terms of its total direct and indirect reverse dependencies.  In current
parlance, it sits C<high upstream on the CPAN river.>  Hence, it ought to work
correctly and be installable on all operating systems where Perl is well
supported.

However, as of early April 2018, FCR version 0.40 wass failing to pass its tests against either
Perl 5.26 or Perl 5 blead on important operating systems including Windows,
FreeBSD and NetBSD
(L<http://fast-matrix.cpantesters.org/?dist=File-Copy-Recursive%200.40>).  As
a consequence, CPAN installers such as F<cpan> and F<cpanm> were failing to
install it (unless one resorted to the C<--force> option).  This prevented
distributions dependent (directly or indirectly) on FCR from being installed
as well.

Some patches had been provided to the L<FCR bug
tracker|https://rt.cpan.org/Dist/Display.html?Name=File-Copy-Recursive> for
this problem.  However, as late as April 18 2018 those patches had not yet
been applied.  This posed a critical problem for the ability to assess the
impact of the soon-to-be-released perl-5.28.0 on CPAN distributions (the
so-called "Blead Breaks CPAN" ("BBC") problem) on platforms other than Linux.

F<File::Copy::Recursive::Reduced> (hereinafter referred to as B<FCR2>) is
intended to provide a minimal subset of FCR's functionality -- just enough to
get the Perl toolchain working on the platforms where FCR is currently
failing.  Functions will be added to FCR2 only insofar as investigation shows
that they can replace usage of FCR functions in toolchain and other heavily
used modules.  No attempt will be made to reproduce all the functionality
currently provided or claimed to be provided by FCR.

On April 19 2018, FCR's author, Daniel Muey, released version 0.41 to CPAN.
This version included a patch submitted by Tom Hukins which corrected the
problem addressed by FCR2.  FCR once again built and tested correctly on
FreeBSD.  That meant that its 6000-plus reverse dependencies can once again be
reached by F<cpan> and other installers.  That in turn means that we can
conduct exhaustive BBC investigations on FreeBSD and other platforms.

With that correction in FCR, the original rationale for FCR2 has been
superseded.  I will continue to maintain the code and respond to bug reports,
but am suspending active development.  I now deem FCR2 feature-complete.

=head1 SUBROUTINES

The current version of FCR2 provides three exportable and publicly supported
subroutines partially equivalent to the similarly named subroutines exported
by FCR.

=head2 C<fcopy()>

=over 4

=item * Purpose

A stripped-down replacement for C<File::Copy::Recursive::fcopy()>.

Copies a file to a new location, recursively creating directories as needed.
Does not copy directories.  Unlike C<File::Copy::copy()>, C<fcopy()> attempts
to preserve the mode of the original file.

=item * Arguments

    fcopy($orig, $new) or die $!;

List of two required arguments:

=over 4

=item * Absolute path to the file being copied; and

=item * Absolute path to the location to which the file is being copied.

=back

Four cases should be noted:

=over 4

=item 1 Create copy within same directory but new basename

    fcopy('/path/to/filename', '/path/to/newfile');

The second argument must be the absolute path to the new file.  (Otherwise
the file will be created in the current working directory, which is almost
certainly what you do not want.)

=item 2 Create copy within different, already B<existing> directory, same basename

    fcopy('/path/to/filename', '/path/to/existing/directory');

The second argument can be merely the path to the existing directory; will
create F</path/to/existing/directory/filename>.

=item 3 Create copy within different, not yet existing directory, same basename

    fcopy('/path/to/filename', '/path/not/yet/existing/directory/filename');

The second argument will be interpreted as the complete path to the newly
created file.  The basename must be included even if it is the same as in the
first argument.  Will create F</path/not/yet/existing/directory/filename>.

=item 4 Create copy within different, not yet existing directory, different basename

    fcopy('/path/to/filename', '/path/not/yet/existing/directory/newfile');

The second argument will be interpreted as the complete path to the newly
created file.  Will create F</path/not/yet/existing/directory/newfile>.

=back

=item * Return Value

Returns C<1> upon success; C<0> upon failure.  Returns an undefined value if,
for example, function cannot validate arguments.

=item * Comment

Since C<fcopy()> internally uses C<File::Copy::copy()> to perform the copying,
the arguments are subject to the same qualifications as that function's
arguments.  Call F<perldoc File::Copy> for discussion of those arguments.

=back

=cut

sub fcopy {
    return unless @_ == 2;
    my ($from, $to) = @_;
    #return unless _samecheck($from, $to);
    return unless _basic_samecheck($from, $to);

    # TODO:  Explore whether we should check (-e $from) here.
    # If we don't have a starting point, it shouldn't make any sense to go
    # farther.

    return unless _dev_ino_check($from, $to);

    return _fcopy($from, $to);
}

sub _fcopy {
    my ($from, $to) = @_;
    my ( $volm, $path ) = File::Spec->splitpath($to);

    # TODO: Explore whether it's possible for $path to be Perl-false in
    # following line.  If not, remove.
    if ( $path && !-d $path ) {
        pathmk(File::Spec->catpath($volm, $path, ''));
    }

    if ( -l $from && $CopyLink ) {
        my $target = readlink( $from );
        # FCR: mass-untaint is OK since we have to allow what the file system does
        ($target) = $target =~ m/(.*)/;
        warn "Copying a symlink ($from) whose target does not exist"
          if !-e $target;
        my $new = $to;
        unlink $new if -l $new;
        symlink( $target, $new ) or return;
    }
    elsif (-d $from && -f $to) { return; }
    else {
        copy($from, $to) or return;

        my @base_file = File::Spec->splitpath( $from );
        my $mode_trg = -d $to ? File::Spec->catfile( $to, $base_file[$#base_file] ) : $to;

        chmod scalar((stat($from))[2]), $mode_trg;
    }
    return 1;
}

sub pathmk {
    my ( $vol, $dir, $file ) = File::Spec->splitpath( shift() );

    # TODO: Exploration whether $dir can be undef at this point.
    # If possible, then we should probably return immediately.
    if ( defined($dir) ) {
        my (@dirs) = File::Spec->splitdir($dir);

        for ( my $i = 0; $i < scalar(@dirs); $i++ ) {
            my $newdir = File::Spec->catdir( @dirs[ 0 .. $i ] );
            my $newpth = File::Spec->catpath( $vol, $newdir, "" );
            mkdir( $newpth );
            return unless -d $newpth;
        }
    }

    # TODO: Exploration whether $file can be undef at this point.
    # If possible, then we should probably return immediately.
    if ( defined($file) ) {
        my $newpth = File::Spec->catpath( $vol, $dir, $file );
        mkdir( $newpth );
        return unless -d $newpth;
    }

    return 1;
}


=head2 C<dircopy()>

=over 4

=item * Purpose

A stripped-down replacement for C<File::Copy::Recursive::dircopy()>.

Given the path to the directory specified by the first argument, the function
copies all of the files and directories beneath it to the directory specified
by the second argument.

=item * Arguments

    my $count = dircopy($orig, $new);
    warn "dircopy() returned undefined value" unless defined $count;

=item * Return Value

Upon completion, returns the count of directories and files created -- which
might be C<0>.

Should the function not complete (but not C<die>), an undefined value will be
returned.  That generally indicates problems with argument validation.  This
approach is taken for consistency with C<File::Copy::Recursive::dircopy()>.

In list context the return value is a one-item list holding the same value as
returned in scalar context.  The three-item list return value of
C<File::Copy::Recursive::dircopy()> is not supported.

=item * Restrictions

None of C<File::Copy::Recursive::dircopy>'s bells and whistles.  No guaranteed
preservation of file or directory modes.  No restriction on maximum depth.  No
nothing; this is fine-tuned to the needs of Perl toolchain modules and their
test suites.

=back

=cut

sub dircopy {

    # I'm not supporting the buffer limitation, at this point I can insert a
    # check for the correct number of arguments:  2
    # FCR2 dircopy does not support buffer limit as third argument

    return unless @_ == 2;

    # Check the definedness and string inequality of the arguments now;
    # Failure to do it now means that if $_[0] is not defined, you'll get an
    # uninitalized value warning in the first line that calls 'substr' below.

    return unless _basic_samecheck(@_);

    # See local file globstar-investigation.pl
    # What the block above does is to trim the 'from' argument so that, if user
    # said 'dircopy(/path/to/directory/*, /path/to/copy)', the first argument
    # is effectively reduced to '/path/to/directory/' but inside $globstar is
    # set to true.  Have to see what impact of $globstar true is.

    return _dircopy(@_);
}

sub _dircopy {
    my $globstar = 0;
    my $_zero    = $_[0];
    my $_one     = $_[1];
    if ( substr( $_zero, ( 1 * -1 ), 1 ) eq '*' ) {
        $globstar = 1;
        $_zero = substr( $_zero, 0, ( length($_zero) - 1 ) );
    }

    # Note also that, in the above, $_[0] and $_[1], while assigned to
    # variables, are not shifted-in.  Hence they retain their original values.
    # TODO: Investigate whether replacing $_[1] from this point forward with a
    # 'my' variable would be harmful.

    # Both arguments must now be defined (though not necessarily true -- yet);
    # they can't be equal; they can't be "dev-ino" equal on non-Win32 systems.
    # Verify that.

    return unless _dev_ino_check( $_zero, $_[1] );

    if ( !-d $_zero || ( -e $_[1] && !-d $_[1] ) ) {
        $! = 20;
        return;
    }

    # If the second argument is not an already existing directory,
    # then, create that directory now (the top-level 'to').

    if ( !-d $_[1] ) {
        pathmk( $_[1] ) or return;
    }
    # If the second argument is an existing directory ...
    # ... $globstar false is the typical case, i.e., no '/*' at end of 2nd argument

    my $baseend = $_one;
    my $level   = 0;
    my $filen   = 0;
    my $dirn    = 0;

    my $recurs;    #must be my()ed before sub {} since it calls itself
    $recurs = sub {
        my ( $str, $end ) = @_;
        $filen++ if $end eq $baseend;
        $dirn++  if $end eq $baseend;

        # On each pass of the recursive coderef, create the directory in the
        # 2nd argument or return (undef) if that does not succeed

        mkdir( $end ) or return if !-d $end;
        $level++;

        opendir( my $str_dh, $str ) or return;
        my @entities = grep( $_ ne '.' && $_ ne '..', readdir($str_dh) );
        closedir $str_dh;

        for my $entity (@entities) {
            my ($entity_ut) = $entity =~ m{ (.*) }xms;
            my $from = File::Spec->catfile( $str, $entity_ut );
            my $to = File::Spec->catfile( $end, $entity_ut );
            if ( -l $from && $CopyLink ) {
                my $target = readlink($from);
                # mass-untaint is OK since we have to allow what the file system does
                ($target) = $target =~ m/(.*)/;
                warn "Copying a symlink ($from) whose target does not exist"
                  if !-e $target;
                unlink $to if -l $to;
                symlink( $target, $to ) or return;
            }
            elsif ( -d $from ) {
                my $rc;
                $rc = $recurs->( $from, $to );
                return unless $rc;
                $filen++;
                $dirn++;
            }
            else {
                fcopy( $from, $to ) or return;
                $filen++;
            }
        } # End 'for' loop around @entities
        $level--;
        1;

    }; # END definition of $recurs

    $recurs->( $_zero, $_one ) or return;
    return $filen;
}

sub _basic_samecheck {
    my ($from, $to) = @_;
    return if !defined $from || !defined $to;
    return if $from eq $to;
    return 1;
}

sub _dev_ino_check {
    my ($from, $to) = @_;
    return 1 if $^O eq 'MSWin32';

    # perldoc perlport: "(Win32) "dev" and "ino" are not meaningful."
    # Will probably have to add restrictions for VMS and other OSes.
    my $one = join( '-', ( stat $from )[ 0, 1 ] ) || '';
    my $two = join( '-', ( stat $to   )[ 0, 1 ] ) || '';
    if ( $one and $one eq $two ) {
        warn "$from and $to are identical";
        return;
    }
    return 1;
}

=head2 C<rcopy()>

=over 4

=item * Purpose

A stripped-down replacement for C<File::Copy::Recursive::rcopy()>.  As is the
case with that FCR function, C<rcopy()> is more or less a wrapper around
C<fcopy()> or C<dircopy()>, depending on the nature of the first argument.

=item * Arguments

    rcopy($orig, $new) or die $!;

List of two required arguments:

=over 4

=item * Absolute path to the entity (file or directory) being copied; and

=item * Absolute path to the location to which the entity is being copied.

=back

=item * Return Value

Returns C<1> upon success; C<0> upon failure.  Returns an undefined value if,
for example, function cannot validate arguments.

=item * Comment

Please read the documentation for C<fcopy()> or C<dircopy()>, depending on the
nature of the first argument.

=back

=cut

sub rcopy {
    return unless @_ == 2;
    my ($from, $to) = @_;
    return unless _basic_samecheck($from, $to);

    # TODO:  Explore whether we should check (-e $from) here.
    # If we don't have a starting point, it shouldn't make any sense to go
    # farther.

    return unless _dev_ino_check($from, $to);

    # symlinks not yet supported
    #return if -l $_[0];
    goto &fcopy if -l $_[0] && $CopyLink;

    goto &_dircopy if -d $_[0] || substr( $_[0], ( 1 * -1 ), 1 ) eq '*';
    goto &_fcopy;
}


=head2 File::Copy::Recursive Subroutines Not Supported in File::Copy::Recursive::Reduced

As of the current version, FCR2 has no publicly documented, exportable subroutines equivalent
to the following FCR exportable subroutines:

    rcopy_glob
    fmove
    rmove
    rmove_glob
    dirmove
    pathempty
    pathrm
    pathrmdir

Consideration is being given to supporting C<rcopy()>.

=head1 BUGS AND SUPPORT

Please report any bugs by mail to C<bug-File-Copy-Recursive-Reduced@rt.cpan.org>
or through the web interface at L<http://rt.cpan.org>.

=head1 ACKNOWLEDGEMENTS

Notwithstanding the fact that this distribution is being released to address
certain problems in File-Copy-Recursive, credit must be given to FCR author
L<Daniel Muey|http://www.cpan.org/authors/id/D/DM/DMUEY/> for ingenious
conception and execution.  The implementation of the subroutines provided by
FCR2 follows that found in FCR to a significant extent.

Thanks also to Tom Hukins for supplying the patch which corrects FCR's
problems and which has been incorporated into FCR2 as well.

Thanks to Håkon Hægland for paying attention to how this library performs on
Windows and other platforms to which the author does not have access.

=head1 AUTHOR

    James E Keenan
    CPAN ID: JKEENAN
    jkeenan@cpan.org
    http://thenceforward.net/perl

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

Copyright James E Keenan 2018-2023.  All rights reserved.

=head1 SEE ALSO

perl(1). File::Copy::Recursive(3).

=cut

1;

__END__

